File: lec1/examples/xv6-c/forkexec.c

#include "kernel/types.h"
#include "user/user.h"

// forkexec.c: fork then exec

int
main()
{
  int pid, status;

  pid = fork();
  if(pid == 0){
    char *argv[] = { "echo", "THIS", "IS", "ECHO", 0 };
    exec("echo", argv);
    printf("exec failed!\n");
    exit(1);
  } else {
    printf("parent waiting\n");
    wait(&status);
    printf("the child exited with status %d\n", status);
  }

  exit(0);
}

File: lec1/examples/linux-rust/src/forkexec.rs
use std::process::Command;
use std::os::unix::process::ExitStatusExt;

fn main() {
    let mut child = Command::new("echo")
        .arg("THIS")
        .arg("IS")
        .arg("ECHO")
        .spawn()
        .expect("failed to execute child process");
    
    println!("parent waiting");
    let status = child.wait().expect("failed to wait on child");

    if let Some(exit_code) = status.code() {
        println!("the child exited with status {}", exit_code);
    } else if let Some(signal) = status.signal() {
        println!("the child was terminated by signal {}", signal);
    }
}